### ISSUE TITLE
[FR] Add "Retry" interval (mins)

### ISSUE THREAD

**Issue by JackDandy** [sabnzbd/sabnzbd#326]
Feature request: Add ***"Retry Interval"*** (mins).

Feature description: Have Sabnzbd wait at least nn configurable minutes in between each retry attempt set in the existing (as of 0.7.xx) ***"Maximum retries"*** setting in the *config/switches/Server* section.

Rationale, I often find myself calling up Sabnzbd to be greeted with failed jobs. I know why this occurs and a click on the retry link _does_ in 99.99% of cases, resolve the fail into a success.  Adding a delay would remove the need to manually intervene.

tia.

**Comment by shypike** [comment]
The "retry" isn't there for the purpose you think it's for.
The maximum retries is about how often SABnzbd retries a server when it times out. This is at the article level, not the NZB level.
Is this about your Usenet provider being slower than the indexer site where you get your NZB-s from?
If so, this will not help.
For a future release we're looking at an option to delay all NZB's with a configurable amount of minutes.
I think that solves your problem better.

**Comment by JackDandy** [comment]
@shypike Yes, you have it, "NZB level" is exactly what I am talking about, and not the existing article level.

Okay, as it is I have already experimented with a global delay on all nzbs.  In practice, simply setting ***all*** NZBs to wait for 10mins was way too cumbersome because some could actually complete quicker and there was no way to know if an item was one that could complete quickly or not.

Exactly what I was asking about is the way to go here, that is an NZB level retry and and a related interval timer.  This flow ensures that transfers that can complete, do so in the shortest time, where others that fail are handled by the interval flow.

**Comment by Safihre** [comment]
I don't get it, you say that you already implemented a standard waiting time of 10min and that it did not work. So why would a configurable value work? It's still static and the same for all NZB's. 
Also, what makes them fail in the first place? Not on your server yet?

**Comment by JackDandy** [comment]
Okay, it would work because the configurable interval value would ***only*** be used when a failure occurs.
1) Transfers that ***can*** complete immediately, will do as is the process right now.
2) ***Only*** transfers that fail will use the additional setting.
Therefore, this is operationally different to when I applied the static value across all transfers.

**Comment by Safihre** [comment]
So you would want that for any time of fail reason? 

@shypike that does seem like an interesting feature?

**Comment by shypike** [comment]
The trouble is, as usual, the correct algorithm.
When trigger an automatic retry? For each failed download? Do you know any criteria for distinguishing a removed post from a post that isn't complete yet? Especially since removed posts are the most likely thing nowadays.

**Comment by thezoggy** [comment]
i think what jackdandy is asking for is delayed retries. right now sab just retries when the article is missing right then for x times. i think he just want to delay (config) when it is retried.. that way you dont have to just assume there will be missing articles and delay up front for no reason.

**Comment by JackDandy** [comment]
@thezoggy has understood me exactly. :)

> if the ending result is the fact it failed due to not enough data.. queue it up again after x mins.

Thanks guys for considering.

**Comment by JackDandy** [comment]
Is this [FR] not worthy of a pretty CYAN label ? :(

**Comment by albino1** [comment]
There's been a lot of posts in the last few weeks on the forums and indexers with people having severe propagation issues, which reminded me of this feature request.

If the indexer is pulling headers from Server A and sends the NZB before Server B has fully received the files the result is that the user sees a failed file with missing blocks, even though they're not missing in the traditional sense, so much as just haven't arrived yet.

An auto-retry on the missing blocks after an interval of time would help alleviate a lot of those problems I suspect.  Alternatively, even something as simple as a way to introduce an intentional propagation delay (like NZBGet has) would at least give the user the option of off-setting the time if they feel it's necessary.

**Comment by JackDandy** [comment]
Yes, the propagation symptom that you raise would be mitigated by my proposal, the logic I suggested should *only* comes into effect when there is an actually fail (like propagation).  
  
However, I feel adding a blanket "intentional propagation delay" on all transfers is actually a step in the wrong direction as it often not needed (i.e. an older release item has no need for a delay).  
  
All things considered, I still believe my suggestion is the most complete solution to the symptom while also covering non-symptomatic downloads and I would wager probably the simplest to implement too.

**Comment by JackDandy** [comment]
@Safihre 

Thank you, unfortunately your PR misses the point.  Users can already add a delay at the nzb site level therefore that feature has been used to find that this method is close to annoying in actual real time use...

Again, for example, say the user snatches any release that is OLDER than 10 mins or whatever, then your option now waits "propagation time" setting before starting - regardless - even if there is no chance of a propagation issue, this is bad.

As explained, there is a simple and more targeted way to approach propagation issues.

**Comment by Safihre** [comment]
No, it is **not** going to delay everything, it looks at the age of the post on the sever and only delays posts that are too young, until they reach the set age. 
 
So in your case if you set on both the indexer 10min and in sabnzbd also 10min, it will start immediately. 
How does this not resolve your problem?

**Comment by JackDandy** [comment]
Ah ok, but there may be no reason to delay. To delay *when* there is an actual error targets the actual reason why the delay was requested as a feature in the first place. So for example, if server with priority 1 can fullfil but is less than 10 mins, then you add a delay even though it can fulfil, if it fails and then server two is tested, then it may also fail or may pass. You see what I mean? I think you need to add the delay after an actual error has triggered.

**Comment by Safihre** [comment]
I see what you mean, and I tried your suggestion of repeat loop, but it's not as easy to implement as you expect, I tried. 
The added bonus is that this Propagation delay feature is easier for other users to understand its purpose and use. It's also something that is available in other usenet software under the same name, so again clearer to users. 
I also tried implementing a per-server setting, but then it would be very hard to display a status message in the queue as we can do with a general setting. So users might think sabnzbd is being stupid and not downloading, while actually it's waiting. 
As you can see we have to make a compromise :) In the end it might indeed cause some jobs to maybe wait 5min too much, but is that so bad? Don't forget all the time that's lost in the par2 checks in your solution, on may systems this is quite time consuming.

**Comment by JackDandy** [comment]
The wait is annoying depending on delay time. For example, a 15 min propagation delay added to a RSS sourced album that takes 5 seconds to download and process is where the annoyance kicks in. 

However, your solution does seem better than what is currently available so that's progress.

I would expect a thread looping trough all servers and a decision on what to do depending on the result if release failed or passed, the same flow that presents the result to the ui - so you're right, I expect it to not be (too much of) a hassle to implement ;)

**Comment by thezoggy** [comment]
my solution for this would be to analyze the nzb when its being added to the queue, see its age... compare that to the delay value.. then set the amount to __pause__ = `delay_value - nzb_age`.

delay value = 15min
nzb age = 11min
nzb when added has a pause value of 4mins when added.. 

that way people could still just tell it to resume if they want to override the delay logic..

**Comment by albino1** [comment]
I was going to request a way to override the delay manually, but zoggy's solution of a pause value makes a lot of sense and would also allow users to override it by un-pausing at the same time.

**Comment by JackDandy** [comment]
Sorry but nope, that won't do either because that (like the newly added solution in play) does not add the delay in response to **ONLY** the actual reason that it is needed... which is, **when a complete failure occurs from all servers** (likely due to missing articles, which in turn is likely due to propagation delay), everyone keeps missing the point in my simplicity, lol.

**Comment by JackDandy** [comment]
@thezoggy @albino1 Sorry but nope, that won't do either because that (like the newly added solution in play) is not a set of actions that **PURELY TARGETS** the actual reason that it is needed... which is, **when a complete failure occurs from all servers** (likely due to missing articles, which in turn is likely due to propagation delay), everyone keeps missing the point in my simplicity, lol.

Design for propagation mitigation spanning 30 minutes with 5 min interval retries...
```
let attempts_left = 6
while attempts_left is more than 0
    if is failure to read from (any of available servers) then
       subtract 1 from attempts_left
       wait for 5 mins before trying again
    otherwise everything is great
       exit this while loop
    end if
end while loop

if no attempts_left then
   report(failure was not recovered by delay method)
end if
```
**No settings required for this feature so user confusion is avoided with this improved design**.
(my settings idea from sept 2015 is replaced with this easier flow to nuke possible user confusion)

**Comment by thezoggy** [comment]
while i understand your wanting a check nzb then decide.. but this is very costly.. especially for those that have lots of servers (i have 7). thus i'd rather have a min-age for a nzb set before trying.. while this isnt a sure fire fix to propagation delays as that can be random.. it will solve the dilemma of adding a nzb right when it gets posted because i see it (announce channels/whatever) and having to sit on it for any length of time. thus if i had a min-age of 15mins set.. then i reduce the likely hood of starting a nzb prematurely before its propagated.

**Comment by JackDandy** [comment]
I am not sure what is costly given headers can be checked before actual "transmission", personally, I have no cost in letting sab try and then fail the nzb, actually, this is the current behavior anyway.  However, your idea is certainly one to include with the proposed flow to make the whole even more robust for both your and my type of use cases.

**Comment by Safihre** [comment]
@JackDandy I hope you can be happy with what I just put in the new pull request #544.
You can set a delay and if you want it to start right away you set it to force.

**Comment by JackDandy** [comment]
@Safihre, Awesome, I thought you had moved onto the next set of code and that this idea lost hope. Thanks for taking _yet another_ look, if this kills prop issues once and for all then it will be a win that i'm sure many will enjoy :) :+1:

**Comment by JackDandy** [comment]
@Safihre, Great, I thought you had moved onto the next set of code. Thanks for taking _yet another_ look, if this finally kills prop issues then it will be a win, it does read as though the idea from @thezoggy was added not mine of a call to action in the event of failure (unless I'm mistaken), so it's something he can be happy with. But hey, I appreciate it all takes effort, so again, thanks.

**Comment by JackDandy** [comment]
@Safihre, Great, I thought you moved onto other parts of code, thanks for taking _yet another_ look. It will be a win if this finally fixes prop issues, it does read as though the idea from @thezoggy was added and not a call to act to actual failure events (or am I mistaken?). But hey, I appreciate it all takes effort, so again, thanks.
